vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -212,23 +212,25 @@
         if (isEmpty()) {
             return Option.none();
         } else {
-            final Traversable<?> objects = isTraversableAgain() ? this : toStream();
-            final Object o = objects.head();
-            if (o instanceof Number) {
-                final Traversable<Number> numbers = (Traversable<Number>) objects;
-                final double d;
-                if (o instanceof Integer || o instanceof Long || o instanceof Byte || o instanceof BigInteger || o instanceof Short) {
-                    d = numbers.toJavaStream()
-                            .mapToLong(Number::longValue)
-                            .average()
-                            .getAsDouble();
+            final T head;
+            final java.util.stream.Stream<T> elements;
+            if (isTraversableAgain()) {
+                head = head();
+                elements = toJavaStream();
+            } else {
+                final Iterator<T> iterator = iterator();
+                head = iterator.next();
+                elements = Iterator.of(head).concat(iterator).toJavaStream();
+            }
+            if (head instanceof Number) {
+                final java.util.stream.Stream<Number> numbers = (java.util.stream.Stream<Number>) elements;
+                final double avg;
+                if (head instanceof Integer || head instanceof Long || head instanceof Byte || head instanceof BigInteger || head instanceof Short) {
+                    avg = numbers.mapToLong(Number::longValue).average().getAsDouble();
                 } else {
-                    d = numbers.toJavaStream()
```

Comment:
An alternative approach would be to define the translation strategy instead of having a long predicate like this. 

Option 1:
```
// top of class
private static Set<Class> longable = Set.of(Integer.class,Long.class,Byte.class,BigInteger.class, Short.class);
if(longable.contains(head))
 avg = numbers.maptToLong(Number::longValue).average().getAsDouble();
```

Option 2: expressly define the function by type:
```
Map<Class<? extends Number>,Function<Stream<Number,? extends Number>> funcs = Map.of(Integer.class,Number::longValue,Byte.class,Number::longValue,Double.class,Number::doubleValue)

Function<Stream<Number>,Stream<Number>> f = funcs.get(head.getClass()).orElseThrow(RuntimeException::new);
number = f.apply(numbers);
```

Option 3 is to recognize the you can handle the simpler case more tersely:
`if(head instanceof Double || head instance of Float)`

Reply:
The if check isn't a problem. Here is a similar Scala code, taken from [NonFatal.scala](https://github.com/scala/scala/blob/v2.12.1/src/library/scala/util/control/NonFatal.scala):

<img width="882" alt="screen shot 2017-09-12 at 20 14 14" src="https://user-images.githubusercontent.com/743833/30341670-2be121d6-97f7-11e7-98a2-ae59a5c0f24e.png">

Instanceof calls are cheap and take no additional memory like the Map and Function instances you suggested. This is code that is not needed. An if is at the lowest level of the language. I tried to sort the checks in the way that most common primitive types occur first. (However, I have no statistic - it is subjective).

Reply:
Sure. I am not opposed to the `instanceof` checks (per-se). I am saying 

```
if(head instanceof Double || head instanceof Float) {
    avg = numbers.mapToDouble(Number::doubleValue).average().getAsDouble();
else
   avg = number.mapToLong(Number::longValue).average().getAsDouble();
``
exact same outcome, less code.

Reply:
This looks great, sorry - I didn't recognized it 😅

Reply:
@chb0github I thought about the code - I intentionally not checked

```java
if (head instanceof Double || head instanceof Float)
```

because we support all elements that extend Number. For example BigDecimal should be mapped to double instead of long.

However, I'm aware that some types can't be calculated accurately in the double space. I will mention that in the javadoc.